tensorflow/contrib/seq2seq/python/ops/beam_search_decoder.py

```diff
@@ -445,15 +517,55 @@ def _maybe_merge_batch_beams(self, t, s):
       A reshaped version of t with shape `[batch_size, beam_width] + s`.
 
     Raises:
-      TypeError: If `t` is an instance of `TensorArray`.
       ValueError:  If the rank of `t` is not statically known.
     """
+    if isinstance(t, tensor_array_ops.TensorArray):
+      return t
     _check_maybe(t)
     if t.shape.ndims >= 2:
       return self._merge_batch_beams(t, s)
     else:
       return t
 
+  def _maybe_sort_array_beams(self, t, parent_ids, sequence_length):
+    """Maybe sorts beams within a `TensorArray`.
+
+    Args:
+      t: A `TensorArray` of size `max_time` that contains `Tensor`s of shape
+        `[batch_size, beam_width, s]` or `[batch_size * beam_width, s]` where
+        `s` is the depth shape.
+      parent_ids: The parent ids of shape `[max_time, batch_size, beam_width]`.
+      sequence_length: The sequence length of shape `[batch_size, beam_width]`.
+
+    Returns:
+      A `TensorArray` where beams are sorted in each `Tensor` or `t` itself if it
+      is not a `TensorArray` or does not meet shape requirements.
+    """
+    if not isinstance(t, tensor_array_ops.TensorArray):
+      return t
+    if (not t._infer_shape or not t._element_shape
+        or t._element_shape[0].ndims is None or t._element_shape[0].ndims < 1):  # pylint: disable=protected-access
+      tf.logging.warn("The cell state contains a TensorArray that is not "
+                      "amenable to sorting based on the beam search result. "
+                      "For a TensorArray to be sorted, its elements shape "
+                      "must be defined and have at least a rank of 1.")
+      return t
+    shape = t._element_shape[0]  # pylint: disable=protected-access
+    batch_size = tensor_util.constant_value(self._batch_size)
```

Comment:
not sure how i feel about this.  i think if batch_size is statically known, as are the entries in `shape`, its ok to raise a ValueError here.  however, if they're not, you should probably push down an `Assert` on the shape of the output of `ta.stack()` to compare the run-time batch size and rank at that point.